core/arch/arm/pta/benchmark.c

```diff
@@ -0,0 +1,200 @@
+{
+	TEE_Result res;
+
+	if ((TEE_PARAM_TYPE_GET(type, 0) != TEE_PARAM_TYPE_MEMREF_INOUT) ||
+		(TEE_PARAM_TYPE_GET(type, 1) != TEE_PARAM_TYPE_NONE) ||
+		(TEE_PARAM_TYPE_GET(type, 2) != TEE_PARAM_TYPE_NONE) ||
+		(TEE_PARAM_TYPE_GET(type, 3) != TEE_PARAM_TYPE_NONE)) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	mutex_lock(&bench_reg_mu);
+
+	DMSG("Registering timestamp buffer, addr = %p, paddr = %" PRIxPA "\n",
+			p[0].memref.buffer,
+			virt_to_phys(p[0].memref.buffer));
+	bench_ts_global = p[0].memref.buffer;
+
+	/* send back to the optee linux kernel module */
+	res = rpc_reg_global_buf(OPTEE_MSG_RPC_CMD_BENCH_REG_NEW,
+				virt_to_phys((void *)bench_ts_global),
+				sizeof(struct tee_ts_global) +
+				sizeof(struct tee_ts_cpu_buf) * bench_ts_global->cores);
+
+	mutex_unlock(&bench_reg_mu);
+
+	return res;
+}
+
+static TEE_Result get_benchmark_memref(uint32_t type,
+				TEE_Param p[TEE_NUM_PARAMS])
+{
+	if ((TEE_PARAM_TYPE_GET(type, 0) != TEE_PARAM_TYPE_VALUE_OUTPUT) ||
+		(TEE_PARAM_TYPE_GET(type, 1) != TEE_PARAM_TYPE_NONE) ||
+		(TEE_PARAM_TYPE_GET(type, 2) != TEE_PARAM_TYPE_NONE) ||
+		(TEE_PARAM_TYPE_GET(type, 3) != TEE_PARAM_TYPE_NONE)) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	mutex_lock(&bench_reg_mu);
+
+	DMSG("Sending back timestamp buffer paddr = %p\n",
+		(void *)virt_to_phys(bench_ts_global));
+	p[0].value.a = virt_to_phys(bench_ts_global);
+
+	mutex_unlock(&bench_reg_mu);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result unregister_benchmark(uint32_t type,
+				TEE_Param p[TEE_NUM_PARAMS] __unused)
+{
+	TEE_Result res;
+
+	if ((TEE_PARAM_TYPE_GET(type, 0) != TEE_PARAM_TYPE_NONE) ||
+		(TEE_PARAM_TYPE_GET(type, 1) != TEE_PARAM_TYPE_NONE) ||
+		(TEE_PARAM_TYPE_GET(type, 2) != TEE_PARAM_TYPE_NONE) ||
+		(TEE_PARAM_TYPE_GET(type, 3) != TEE_PARAM_TYPE_NONE)) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+	mutex_lock(&bench_reg_mu);
+
+	DMSG("Unregistering benchmark, timestamp buffer paddr = %p\n",
+		(void *)virt_to_phys(bench_ts_global));
+	bench_ts_global = NULL;
+
+	res = rpc_reg_global_buf(OPTEE_MSG_RPC_CMD_BENCH_REG_DEL, 0, 0);
+
+	mutex_unlock(&bench_reg_mu);
+
+	return res;
+}
+
+static TEE_Result invoke_command(void *pSessionContext __unused,
+		uint32_t nCommandID, uint32_t nParamTypes,
+		TEE_Param pParams[TEE_NUM_PARAMS])
+{
+	switch (nCommandID) {
+	case BENCHMARK_CMD_REGISTER_MEMREF:
+		return register_benchmark_memref(nParamTypes, pParams);
+	case BENCHMARK_CMD_GET_MEMREF:
+		return get_benchmark_memref(nParamTypes, pParams);
+	case BENCHMARK_CMD_UNREGISTER:
+		return unregister_benchmark(nParamTypes, pParams);
+	default:
+		break;
+	}
+
+	return TEE_ERROR_BAD_PARAMETERS;
+}
+
+pseudo_ta_register(.uuid = BENCHMARK_UUID, .name = TA_NAME,
+		   .flags = PTA_DEFAULT_FLAGS,
+		   .invoke_command_entry_point = invoke_command);
+
+/* Cycle counter */
+static inline uint64_t read_ccounter(void)
+{
+	uint64_t ccounter = 0;
+#ifdef __aarch64__
+	asm volatile("mrs %0, PMCCNTR_EL0" : "=r"(ccounter));
```

Comment:
Please add a `read_pmccntr_el0()` or similar in `arm64.h` and use that instead, and something similar for the ARM32 version below.